File: Problem-2.java
//https://leetcode.com/problems/minimum-window-substring/
// Time Complexity :O(s*t)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :

//Sliding window
// Your code here along with comments explaining your approach
class Solution {
    public String minWindow(String s, String t) {
        
        if(t==null || t.length() ==0 || t.length() > s.length()) return "";
        
        HashMap<Character , Integer> tMap = new HashMap<>();
        
        HashMap<Character , Integer> sMap = new HashMap<>();
        
        int start =0;
        int min = t.length();
        int count =0;
        String result ="";
        
        for(int i =0 ;i<t.length();i++)
        {
            tMap.put(t.charAt(i) , tMap.getOrDefault(t.charAt(i),0)+1);
        }
        
        for(int end =0 ;end<s.length();end++){
            
            char ch = s.charAt(end);
            sMap.put(ch,sMap.getOrDefault(ch,0)+1);
            
            if(tMap.containsKey(ch) && sMap.getOrDefault(ch ,0) <= tMap.getOrDefault(ch ,0)) count++;
            
            while(start <=end && count == min){
                
                if(result.length() ==0 || end-start+1 < result.length()){
                    result = s.substring(start,end+1);
                }
                
                char sc = s.charAt(start);
               
                sMap.put(sc , sMap.get(sc)-1);
                
                if(sMap.get(sc) ==0) sMap.remove(sc);
                
                if(sMap.getOrDefault(sc,0) < tMap.getOrDefault(sc,0)) count--;
                
                start++;
            }
        }
        
        return result;
    }
}

File: problem2.cpp

// Time Complexity :O(n+m) m = t.length, n = s.length
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
// we use a tMap for strint t and a matched variable
// traverse over s
// while traversing check the character in tMap, 
// if present decrease matched and map value
// if(matched == 0)  we get an answer string
// while increaing start , we increment in map and if map value >=1 we increment matched

class Solution {
public:
    string minWindow(string s, string t) {
        int first = -1,second = -1;
        unordered_map<char,int>tMap;
        int matched = 0;
        for(int i =0;i<t.size();i++) {
            tMap[t[i]]++;
            matched++;
        }
        int start = 0;
        for(int end = 0;end<s.size();end++){
            if(tMap[s[end]]>0){
                matched--;
            }
            tMap[s[end]]--;
            while(matched == 0){
                if(first== -1 ||  (second-first+1 > end-start+1)){
                    first = start;
                    second = end;
                }
                tMap[s[start]]++;
                if(tMap[s[start]]>0) matched++;
                start++;
            }
        }
        return (first==-1)? "":s.substr(first,second-first+1);
    }
};

